File: 백준/Silver/1874. 스택 수열/스택 수열.py
# 1874. 스택 수열 [S2]

n = int(input())
stack, ans, find = [], [], True

# 숫자 1부터 시작
now = 1
for _ in range(n):
    num = int(input())
    # 스택 쌓기 push
    while now <= num:
        stack.append(now)
        ans.append('+')
        now += 1
    # 스택 꺼내기 pop
    if stack[-1] == num:
        stack.pop()
        ans.append('-')
    # 불가능 한 경우
    else:
        find = False

# 정답 출력
if not find:
    print('NO')
else:
    for i in ans:
        print(i)

File: 백준/Silver/1874. 스택 수열/스택 수열.java
// 01874. [S2] 스택 수열
import java.io.*;
import java.util.Stack;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringBuilder sb = new StringBuilder();
    static Stack<Integer> stack = new Stack<>();
    static int num = 1;
    public static void main(String[] args) throws IOException {
        int N = Integer.parseInt(br.readLine());

        for (int n = 0; n < N; n++) {
            int goal = Integer.parseInt(br.readLine());

            while (num <= goal) {
                stack.push(num++);
                sb.append("+\n");
            }

            if (stack.peek() == goal) {
                stack.pop();
                sb.append("-\n");
            } else {
                System.out.println("NO");
                return;
            }
        }
        System.out.println(sb.toString());
        br.close();
    }
}
